File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/camera/FreeCameraOptions.java
package org.maplibre.nativeffi.camera;

import java.util.Objects;
import org.maplibre.nativeffi.geo.Quaternion;
import org.maplibre.nativeffi.geo.Vec3;

/** Mutable descriptor for free-camera position and orientation. */
public final class FreeCameraOptions {
  private Vec3 position;
  private Quaternion orientation;

  public boolean hasPosition() {
    return position != null;
  }

  public Vec3 position() {
    return position;
  }

  public FreeCameraOptions position(Vec3 position) {
    this.position = Objects.requireNonNull(position, "position");
    return this;
  }

  public FreeCameraOptions clearPosition() {
    position = null;
    return this;
  }

  public boolean hasOrientation() {
    return orientation != null;
  }

  public Quaternion orientation() {
    return orientation;
  }

  public FreeCameraOptions orientation(Quaternion orientation) {
    this.orientation = Objects.requireNonNull(orientation, "orientation");
    return this;
  }

  public FreeCameraOptions clearOrientation() {
    orientation = null;
    return this;
  }
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/camera/FreeCameraOptions.kt
package org.maplibre.nativeffi.camera

import org.maplibre.nativeffi.geo.Quaternion
import org.maplibre.nativeffi.geo.Vec3

/**
 * Mutable free-camera descriptor.
 *
 * Compares and hashes by field value; [copy] returns an independent instance. Keep an instance
 * unmodified while it is a key in a hash-based collection.
 */
public class FreeCameraOptions {
  public var position: Vec3? = null

  public var orientation: Quaternion? = null

  /** Returns an independent copy of this descriptor with [block] applied to the copy. */
  public fun copy(block: FreeCameraOptions.() -> Unit = {}): FreeCameraOptions =
    FreeCameraOptions()
      .also {
        it.position = position
        it.orientation = orientation
      }
      .apply(block)

  private val fields: List<Any?>
    get() = listOf(position, orientation)

  override fun equals(other: Any?): Boolean = other is FreeCameraOptions && fields == other.fields

  override fun hashCode(): Int = fields.hashCode()
}
